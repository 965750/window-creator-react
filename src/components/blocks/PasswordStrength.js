import React, { Component } from 'react'
import { FormattedMessage } from 'react-intl'
import PropTypes from 'prop-types'
import styles from './PasswordStrength.module.scss'

class PasswordStrength extends Component {
    state = {
        strength: 0,
        label: null,
    }

    checkStrength = () => {
        let localStrength = 0

        const passwordCheck = {
            lowerAlph: /(?=.*[a-z])/,
            upperAlph: /(?=.*[A-Z])/,
            numericChar: /(?=.*[0-9])/,
            specialChar: /(?=.[!@#%&])/,
            long: /(?=.{12,})/,
        }

        if (passwordCheck.lowerAlph.test(this.props.password)) {
            localStrength += 10
        }
        if (passwordCheck.upperAlph.test(this.props.password)) {
            localStrength += 10
        }
        if (passwordCheck.numericChar.test(this.props.password)) {
            localStrength += 25
        }
        if (passwordCheck.specialChar.test(this.props.password)) {
            localStrength += 30
        }
        if (passwordCheck.long.test(this.props.password)) localStrength += 25

        switch (true) {
            case localStrength < 30:
                this.setState({
                    label: 'Low',
                })

                break
            case localStrength >= 30 && localStrength <= 70:
                this.setState({
                    label: 'Medium',
                })

                break
            case localStrength > 70:
                this.setState({
                    label: 'Good',
                })

                break
            default:
                return
        }

        this.setState({
            strength: localStrength,
        })
    }

    checkedColor = () => {
        switch (this.state.label) {
            case 'Low':
                return 'error'
            case 'Medium':
                return 'warning'
            case 'Good':
                return 'malachite'
            default:
                return 'darkText'
        }
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (prevProps.password !== this.props.password) {
            this.checkStrength()
        }
    }

    render() {
        let label

        if (this.state.label) {
            label = (
                <FormattedMessage
                    id={this.state.label}
                    defaultMessage={this.state.label}
                />
            )
        }

        return (
            <div className={`${styles.bar} relative w-4/9 mb-1`}>
                <p
                    className={`${
                        styles.bar__label
                    } top-10- text-${this.checkedColor()}`}
                >
                    {label}
                </p>
                <div className={`${styles.bar__back} rounded`} />
                <div
                    style={{ width: `${100 - this.state.strength}%` }}
                    className={`${styles.bar__cover} rounded-r absolute bg-white opacity-75 h-full bottom-0 right-0 w-50`}
                />
            </div>
        )
    }
}

PasswordStrength.propTypes = {
    password: PropTypes.string,
}

export default PasswordStrength
