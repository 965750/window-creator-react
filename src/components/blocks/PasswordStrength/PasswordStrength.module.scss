.bar {

  &__label {
    height: 12px;
    line-height: 12px;
    font-size: 11px;
  }

  &__back {
    height: 7px;
    background: linear-gradient(90deg, rgba(255,0,0,1) 0%, rgba(245,255,0,1) 50%, rgba(9,255,0,1) 100%);
  }

  &__cover {
    height: 7px;
  }
}