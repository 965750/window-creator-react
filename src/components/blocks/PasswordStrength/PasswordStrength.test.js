import '@testing-library/jest-dom/extend-expect'

import { applyMiddleware, compose, createStore } from 'redux';
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { render, cleanup } from '@testing-library/react'

import PasswordStrength from './index'
import messages from '../../../messages-i18n'
import thunk from 'redux-thunk';
import React from 'react'

const initialStore = {
  theme: {
    lang: 'pl',
  },
}

const reducer = (state = initialStore, action) => {
  switch (action.type) {
    default:
      return state
  }
}

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
  ),
)

const renderWithRedux = (component) => ({
  ...render(<Provider store={store}>{component}</Provider>),
})

describe('PasswordStrength', () => {
  let wrapper
  const initialProps = {
    password: ''
  }

  const setUp = (props = initialProps) => {
    wrapper = renderWithRedux(
      <IntlProvider
        locale="en"
        messages={messages[store.getState().theme.lang]}
      >
        <PasswordStrength {...props} />
      </IntlProvider>,
    )
  }

  afterEach(cleanup)

  it('should match snapshot', () => {
    setUp()

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should check password strength of 10%', () => {
    setUp({
      password: 'asd'
    })

    expect(wrapper.getByTestId('passwordStrength')).toHaveStyle(`width: 90%`)
    expect(wrapper.queryByText(/Słabe/)).not.toBeNull()
  })

  it('should check password strength of 20%', () => {
    setUp({
      password: 'Asd'
    })

    expect(wrapper.getByTestId('passwordStrength')).toHaveStyle(`width: 80%`)
    expect(wrapper.queryByText(/Słabe/)).not.toBeNull()
  })

  it('should check password strength of 45%', () => {
    setUp({
      password: 'Asd1'
    })

    expect(wrapper.getByTestId('passwordStrength')).toHaveStyle(`width: 55%`)
    expect(wrapper.queryByText(/Średnie/)).not.toBeNull()
  })

  it('should check password strength of 75%', () => {
    setUp({
      password: 'Asd1#'
    })

    expect(wrapper.getByTestId('passwordStrength')).toHaveStyle(`width: 25%`)
    expect(wrapper.queryByText(/Dobre/)).not.toBeNull()
  })

  it('should check password strength of 100%', () => {
    setUp({
      password: '123456789Asd1#'
    })

    expect(wrapper.getByTestId('passwordStrength')).toHaveStyle(`width: 0%`)
    expect(wrapper.queryByText(/Dobre/)).not.toBeNull()
  })
})
