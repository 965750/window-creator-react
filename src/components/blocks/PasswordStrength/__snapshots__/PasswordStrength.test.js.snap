// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordStrength should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="bar relative w-4/9 mb-1"
  >
    <p
      class="bar__label top-10- text-osloGray"
    />
    <div
      class="bar__back rounded"
    />
    <div
      class="bar__cover rounded-r absolute bg-white opacity-75 h-full bottom-0 right-0 w-50"
      data-testid="passwordStrength"
      style="width: 100%;"
    />
  </div>
</DocumentFragment>
`;
